File: jisoo/programmers/level1/키패드_누르기.py
"""
https://school.programmers.co.kr/learn/courses/30/lessons/67256
"""


def getDistance(a, b): #a 다음 누를 위치, b 손가락 위치 a = (1,1) b = (2,0)
  return abs(a[0] - b[0]) + abs(a[1] - b[1])



def solution(numbers, hand):
  answer = ''

  keypad = {
    1: (0, 0), 2: (1, 0), 3: (2, 0),
    4: (0, 1), 5: (1, 1), 6: (2, 1),
    7: (0, 2), 8: (1, 2), 9: (2, 2),
    '*': (0, 3), 0: (1, 3), '#': (2, 3),
  }

  rp = '#'
  lp = '*'

  R = [3, 6, 9]
  L = [1, 4, 7]
  C = [2, 5, 8, 0]

  for i in numbers:
    if i in R:
      answer += 'R'
      rp = i
    if i in L:
      answer += 'L'
      lp = i
    if i in C:
      rd = getDistance(keypad[i], keypad[rp]) # i: 5 rp: 3
      ld = getDistance(keypad[i], keypad[lp])

      if rd > ld:  # 양수면 R이 더큰것 L로 누름
        answer += "L"
        lp = i
      elif rd < ld:
        answer += "R"
        rp = i
      else:
        if hand == 'right':
          answer += "R"
          rp = i
        else:
          answer += "L"
          lp = i

  return answer

File: seongwon/LV1/키패드누르기.js
function solution(numbers, hand) {
  const obj = {
    1: [0, 0],
    2: [0, 1],
    3: [0, 2],
    4: [1, 0],
    5: [1, 1],
    6: [1, 2],
    7: [2, 0],
    8: [2, 1],
    9: [2, 2],
    "*": [3, 0],
    0: [3, 1],
    "#": [3, 2],
  };
  let left = "*";
  let right = "#";
  let result = "";
  for (let num of numbers) {
    if (num === 1 || num === 4 || num === 7) {
      result += "L";
      left = String(num);
    } else if (num === 3 || num === 6 || num === 9) {
      result += "R";
      right = String(num);
    } else {
      let [L, R] = getDistance(obj[left], obj[right], obj[num]);
      if (L < R) {
        result += "L";
        left = String(num);
      } else if (R < L) {
        result += "R";
        right = String(num);
      } else {
        if (hand === "right") {
          result += "R";
          right = String(num);
        } else {
          result += "L";
          left = String(num);
        }
      }
    }
  }
  return result;
}
function getDistance(L, R, curP) {
  let fromL = Math.abs(curP[0] - L[0]) + Math.abs(curP[1] - L[1]);
  let fromR = Math.abs(curP[0] - R[0]) + Math.abs(curP[1] - R[1]);
  return [fromL, fromR];
}
console.log(solution([1, 3, 4, 5, 8, 2, 1, 4, 5, 9, 5], "right"));
